src/libstd/thread/local.rs

```diff
@@ -364,54 +401,79 @@ pub mod fast {
     impl<T> Key<T> {
         pub const fn new() -> Key<T> {
             Key {
-                inner: UnsafeCell::new(None),
-                dtor_registered: Cell::new(false),
-                dtor_running: Cell::new(false)
+                inner: LazyKeyInner::new(),
+                dtor_state: Cell::new(DtorState::Unregistered),
             }
         }
 
-        pub unsafe fn get(&self) -> Option<&'static UnsafeCell<Option<T>>> {
-            if mem::needs_drop::<T>() && self.dtor_running.get() {
-                return None
+        pub unsafe fn get<F: FnOnce() -> T>(&self, init: F) -> Option<&'static T> {
+            match self.inner.get() {
+                Some(val) => Some(val),
+                None => self.try_initialize(init),
```

Comment:
So looking at this again, I'm just trying to follow the branching here and I'm thinking now, would this produce the same codegen?

```rust
match self.inner.get() {
    Some(val) => Some(val),
    None => {
        if mem::needs_drop::<T>() {
            self.register_dtor()?;
        }
        Some(self.inner.initialize(init))
    }
}
```

where `register_dtor` is basically what `try_initialize_drop` is right now except it's missing the last `self.inner.initialize`.

I'd want to be careful to not tamper with the codegen too much though. Is `#[cold]` on `try_initialize` needed for that? If so, then could `try_initialize` be updated to have a similar structure to the `match` I'm thinking of?

Reply:
AFAICT `self.inner.initialize` _has_ to be behind a `#[inline(never)]` function or else there's two `tlv_get_addr` calls on macos. `#[cold]` is optional, but seems to generate better asm IMO.

Putting the `needs_drop` check behind an `inline(never)` like so:
```rust
#[inline(never)]
#[cold]
unsafe fn try_initialize<F: FnOnce() -> T>(&self, init: F) -> Option<&'static T> {
    if !mem::needs_drop::<T>() || self.try_register_dtor() {
        Some(self.inner.initialize(init))
    } else {
        None
    }
}
```

causes unwrapping code to be generated for `LocalKey::with`, even for cases which always return `Some` (`needs_drop::<T>() == false`). I'll poke a little more at this on the weekend, but it is tricky to merge drop/nodrop paths together without hurting one or the other.

Reply:
Ok I think `#[cold]` is definitely an easy win (and one we should do here).

If `#[inline(never)]` is only on `try_register_dtor` (this hypothetical function) I think that the codegen should be basically the same as this PR currently is, meaning no unwrapping code but still two accesses perhaps?

Reply:
The generated code does come out to be the same! I think that's a pretty good win.

Moreover, I was wrong about needing `#[inline(never)]` anywhere. Only using `#[cold]` and only on `try_initialize` gives pretty equivalent results. Sometimes two accesses (`try_initialize` is inlined), sometimes 1 (`try_initialize` is not inlined).

@alexcrichton Thanks for help and time spent on this PR! I pushed up the inlining/refactor changes.